Fixed a few issues with uppy/core types

DJWassink opened the pull request:
Was actually going to fix #954, but it seems like a fix for that one is already merged :) I guess that issue could be closed now. Anyhow I found some more typing things that could be improved.

Sweet job on the latest release! Better tree shaking / lerna + ts types, what a treat 😍

arturi commented:
Sure. Thank you! ❤️

DJWassink commented:
Just now added some generic's to the files and splitted the upload result files in two, one with an error string and the other with an uploadurl, can be handy when you use getstate to get a list to let typescript differentiate between a failed file and a uploaded file.

I'm not quite happy with the generics tho, I tried to set the generic on the main Uppy object but that introduces a whole lot of other issues, especially within addons.

Anyhow feel free to leave feedback on any changes if changes are required :)

arturi commented:
This looks good to me! @goto-bus-stop ?

goto-bus-stop commented on packages/@uppy/core/types/index.d.ts:
```diff
@@ -56,29 +70,47 @@ export interface Store {
 }
 
+interface LocaleObject {
+  [key: string]: string | LocaleObject;
```
A locale object can only have a `strings` and `pluralize` key, I think it should be:

```ts
interface LocaleStrings {
  [key: string]: string | LocaleStrings;
}
interface Locale {
  strings: LocaleStrings;
  pluralize: (n: number) => number;
}
```

DJWassink commented:
Can't say I remember 100% what the issue was (been on holiday and abandoned all work related stuff ;) ) but at some point I was adding generics to all kind of different places and if I remember correctly some plugins also needed the generic but typescript couldn't figure out that the plugin was then a part of uppy, so the uppy '.use' method wouldn't accept new plugins anymore (for example the s3-multi-part).

Anyhow because of a fuck-up on my part my laptop is completely wiped so I can't recover my attempt to type uppy itself but honestly I think we should abandon that idea (or some ts wizard has to pass by). At the moment I am cloning the repo again and installing the deps. I will see if I can do something about your feedback and if there is something else I would like to fix. Else I think this PR should be merged for now and I will open another one in the future when needed :)

goto-bus-stop commented:
> [..] some plugins also needed the generic [..]

Oh that makes sense, with plugins having the Uppy instance as a parameter in the constructor and with the plugin-specific `uppy.use()` typings. Thanks for the elaboration!

RIP data, hope you didn't lose anything too important :sweat_smile:

DJWassink commented:
Haha nothing of value has been lost, atleast as far as I know 🤔 

Anyhow I have changed to locale interface thnx to your feedback, should reflection the situation in the docs now :). Also removed the WIP in the title since I'm not going to change a whole lot for now (maybe in the future who knows ;) )

DJWassink commented:
Currently the tests are failing because i also typed that the `addFile` method needs and type and a name. I did this because if I only supply the data property I get the following error: 

```
xxx.ts:67 TypeError: Cannot read property 'split' of null
    at Uppy.addFile (index.js:451)
```

![image](https://user-images.githubusercontent.com/1822404/44343944-9a408b80-a48f-11e8-920d-1c5e42c78681.png)

The following line is the culprit of this error:
https://github.com/transloadit/uppy/blob/f5bde1096cfe4c24c066e8863e11911a190f78bf/packages/%40uppy/core/src/index.js#L409

I guess I should revert the changes made here and uppy core itself should be fixed with a null-check for the fileType (it already has a fallback to noname which it won't reach because of the error)

What do you guys think?

goto-bus-stop commented:
> I guess I should revert the changes made here and uppy core itself should be fixed with a null-check for the fileType (it already has a fallback to noname which it won't reach because of the error)

That sounds like the best approach to me (or mark name/type optional in the AddFileOptions interface)

DJWassink commented:
I have reverted the changes, since options objects already extends a `Partial<UppyFile>` all the other props are already marked optional ;)

goto-bus-stop commented:
Sweet :ok_hand: thanks!
